Framework/DataHandling/src/SaveSESANS.cpp

```diff
@@ -0,0 +1,195 @@
+ */
+void SaveSESANS::init() {
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      "InputWorkspace", "", Kernel::Direction::Input),
+                  "The name of the workspace to save");
+  declareProperty(Kernel::make_unique<API::FileProperty>(
+                      "Filename", "", API::FileProperty::Save, fileExtensions),
+                  "The name to use when saving the file");
+
+  // TODO : find out good descriptions (and validators) for these properties
+  declareProperty("ThetaZMax", -1.0, Kernel::Direction::Input);
+  declareProperty("ThetaZMax_unit", "radians", Kernel::Direction::Input);
+  declareProperty("ThetaYMax", -1.0, Kernel::Direction::Input);
+  declareProperty("ThetaYMax_unit", "radians", Kernel::Direction::Input);
+  declareProperty("EchoConstant", -1.0, Kernel::Direction::Input);
+
+  declareProperty<std::string>("Orientation", "",
+                               "Orientation of the instrument");
+}
+
+/**
+ * Execute the algorithm
+ */
+void SaveSESANS::exec() {
+  API::MatrixWorkspace_const_sptr ws = getProperty("InputWorkspace");
+
+  // Check workspace has only one spectrum
+  if (ws->getNumberHistograms() != 1) {
+    g_log.error("This algorithm expects a workspace with exactly 1 spectrum");
+    throw std::runtime_error("SaveSESANS passed workspace with incorrect "
+                             "number of spectra, expected 1");
+  }
+
+  std::ofstream outfile;
+  outfile.open(getPropertyValue("Filename"));
+  writeHeaders(outfile, ws);
+  outfile << "\n"
+          << "BEGIN_DATA"
+          << "\n";
+
+  const auto &wavelength = ws->points(0);
+  const auto &yValues = ws->y(0);
+  const auto &eValues = ws->e(0);
+
+  const auto spinEchoLength = calculateSpinEchoLength(wavelength);
+  const auto depolarisation = calculateDepolarisation(yValues, wavelength);
+  const auto error = calculateError(eValues, yValues, wavelength);
+
+  outfile << "SpinEchoLength Depolarisation Depolarisation_error Wavelength\n";
+  // outfile.precision(5);
+  // outfile << std::fixed;
+
+  for (size_t i = 0; i < spinEchoLength.size(); ++i) {
+    outfile << spinEchoLength[i] << " ";
+    outfile << depolarisation[i] << " ";
+    outfile << error[i] << " ";
+    outfile << wavelength[i] << "\n";
+  }
+
+  outfile.close();
+}
+
+/**
+ * Write header values to the output file
+ * @param outfile ofstream to the output file
+ * @param ws The workspace to save
+ */
+void SaveSESANS::writeHeaders(std::ofstream &outfile,
+                              API::MatrixWorkspace_const_sptr &ws) {
+  const API::Sample &sample = ws->sample();
+
+  writeHeader(outfile, "FileFormatVersion", "1.0");
+  writeHeader(outfile, "DataFileTitle", ws->getTitle());
+  writeHeader(outfile, "Sample", sample.getName());
+  writeHeader(outfile, "Thickness", std::to_string(sample.getThickness()));
+  writeHeader(outfile, "Thickness_unit", "mm");
+  writeHeader(outfile, "Theta_zmax", getPropertyValue("ThetaZMax"));
```

Comment:
You should check these in a `validateInputs` method, as you had suggested in the `TODO` section